### PR TITLE
 Authorizations : cancancan

### PR DESCRIPTION
**Pull request by aqfaridi**
**What this PR does :** 
- Integrated Select2 for dropdown Tag search
- Switched to mysql database from sqlite
- Role based authorization - viewer, blogger, admin

**Dependency :** 
- cancancan gem

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -34,6 +34,8 @@ gem 'acts-as-taggable-on', '~> 3.4'
 gem "paperclip", '~> 4.3.6'
 gem 'bootstrap-tagsinput-rails', '~> 0.4.2.1'
 gem 'ransack', '~> 1.7.0'
+gem 'cancancan', '~> 1.10'
```
Comment (coolRaider) [inline Gemfile:47]: add reason for including gem as comment

File: app/controllers/articles_controller.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@ def new
 
   def edit
     @article = Article.find(params[:id])
-    validation(current_user,@article)  
+    #validation(current_user,@article)
```
Comment (coolRaider) [inline app/controllers/articles_controller.rb]: please remove commented out code

File: app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class RegistrationsController < Devise::RegistrationsController
+# before_action :configure_sign_in_params, only: [:create]
```
Comment (coolRaider) [inline app/controllers/registrations_controller.rb]: is this class really being used?

File: app/models/ability.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      if user.has_role? :blogger
+        can :create, Article
+        can :update, Article do |article|
+          article.try(:user) == user
+        end
+        can :destroy, Article do |article|
+          article.try(:user) == user
+        end
+        can :manage, Comment do |comment|
+          comment.try(:article).try(:user).eql?(user)
```
Comment (coolRaider) [inline app/models/ability.rb:26]: shouldn't this :manage, Comment  apply to normal users too? not just bloggers?
additionally, bloggers should be able to destroy comments on their own blogs from different users.

File: app/views/articles/index.html.haml [hunk 5]
```diff
@@ -15,3 +15,12 @@
 
 %div.thumb
   = render :partial => 'card'
+
+:javascript
+  jQuery( document ).ready(function( $ ) {
```
Comment (coolRaider) [inline app/views/articles/index.html.haml:20]: all javascript should be included in the assets, to keep it unobtrusive, readable and maintainable. Refer:
http://stackoverflow.com/questions/6287440/rails-3-good-rule-of-thumb-for-where-to-put-javascript

File: app/views/devise/registrations/new.html.haml [hunk 6]
```diff
@@ -6,6 +6,7 @@
         = image_tag("default.png", :alt => "profile", :class => 'profile-img')
         = form_for(resource, as: resource_name, url: registration_path(resource_name),:html => { :class => "form-signin" }) do |f|
           = devise_error_messages!
+          = hidden_field_tag "user[roles][]", "viewer"
```
Comment (coolRaider) [inline app/views/devise/registrations/new.html.haml]: **SECURITY ISSUE:** should this really be set here? potentially someone could just use postman and register as admin?

File: app/views/welcome/index.html.haml [hunk 7]
```diff
@@ -2,3 +2,5 @@
 %h1
   %em Welcome to Blog
 %hr/
+
+
```
Comment (coolRaider) [inline app/views/welcome/index.html.haml]: please remove multiple new lines from the EOF.

File: config/database.yml [hunk 8]
```diff
@@ -4,14 +4,32 @@
...
 
 development:
-  <<: *default
-  database: db/development.sqlite3
+  adapter: mysql2
+  encoding: utf8
+  database: blog
+  pool: 20
+  username: root
+  password: root
```
Comment (coolRaider) [inline config/database.yml]: please remove, username and password from here

File: app/models/ability.rb (hunk 4 above)
Comment (aqfaridi) [inline app/models/ability.rb:26]: yes blogger is able to destroy any comment on his article
& normal user is viewer who has read access to article and can create comment

File: app/models/user.rb [hunk 9]
```diff
@@ -38,4 +39,25 @@ class User < ActiveRecord::Base
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable, :confirmable
   has_many :articles
+
+
+  ROLES = %w[:admin :blogger :viewer]
```
Comment (coolRaider) [inline app/models/user.rb:43]: please remove multiple new lines, is this code auto-generated? documentation suggests only adding two lines here, include RoleModel and available roles. refer:
https://github.com/martinrehfeld/role_model

**Comment by coolRaider** [comment]
Please try to create different branches for different features. Select2 and cancancan should be two separate branches. 40 files with changes in one PR are difficult to review.

File: app/models/ability.rb (hunk 4 above)
Comment (coolRaider) [inline app/models/ability.rb:26]: but here you have given ability to edit and delete comment only to blogger for his comments only. normal user should also be able to do that for his own comments. Try creating two users, comment with one on first one's blog, and try to delete the comment from the other user. Is that working?

File: app/views/devise/registrations/new.html.haml (hunk 6 above)
Comment (aqfaridi) [inline app/views/devise/registrations/new.html.haml]: 👍  nice !! bug fixed

File: app/models/ability.rb (hunk 4 above)
Comment (aqfaridi) [inline app/models/ability.rb:26]: 👍 done

**Pull request merged**